chirp/drivers/retevis_ra25.py

```diff
@@ -0,0 +1,1098 @@
+    except Exception as e:
+        raise errors.RadioError(f"Error Entering Program Mode {e}")
+    return True
+
+
+def end_programming(serial):
+    try:
+        rdata = send_cmd(serial, b'END')
+        if rdata != b'\x06':
+            log.error(f'Error {rdata[0]}')
+    except Exception as e:
+        raise errors.RadioError(f"Error {e}")
+
+
+def parse_response(r):
+    if sum(r[:-1]) % 256 != r[-1]:
+        raise errors.RadioError("Data check sum failed")
+    return r[5:-1]
+
+
+def get_ident(serial):
+    try:
+        bs = send_cmd(serial, b'\x02')
+        model = bs[1:8].decode()
+        mode = bs[8]
+        ver = bs[9:13].decode()
+        log.debug(f'Model {model} Ver {ver} Band Mode {mode}')
+        if SUPPORTED_MODEL_VERSIONS.get(model, []):
+            if ver in SUPPORTED_MODEL_VERSIONS[model]:
+                log.info(f'Model {model} Ver: {ver} supported')
+        else:
+            raise errors.RadioError(
+                f'ERROR: Model: {model} Ver: {ver} is not supported!')
+        return bs
+    except Exception as e:
+        raise errors.RadioError(e)
+
+
+def get_bands(mode):
+    idx = int(hex(mode)[2:]) - 1
+    try:
+        valid_ranges = BAND_LIMITS[idx]
+    except IndexError:
+        raise errors.RadioError("Unknown Radio Mode - Can't set Freq. Bands!")
+    return valid_ranges
+
+
+def do_download(radio):
+    try:
+        data = b''
+        # Get the serial port connection
+        serial = radio.pipe
+        status = chirp_common.Status()
+        status.msg = "Connecting to Radio..."
+        radio.status_fn(status)
+        r = start_programming(serial)
+        if not r:
+            raise errors.RadioError('Failed to enter programming mode')
+        ident_bytes = get_ident(serial)
+        data = ident_bytes
+        status.max = sum((t[1] * t[2]) for t in MEMORY_REGIONS if t[3] != 'S')
+        for reg in MEMORY_REGIONS:
+            base, blocks, blk_sz, mod, struc, label = reg
+            if mod != 'S':
+                status.msg = f"Downloading: {label}..."
+                base_int = int(base, 16)
+                addrs = [hex(base_int+i)[2:] for i in range(0, blocks)]
+                for addr in addrs:
+                    cmd = bytes.fromhex('52' + addr.zfill(8))
+                    r = send_cmd(serial, cmd)
+                    if r[0] == 0x57:
+                        data += parse_response(r)
+                        status.cur += len(r[5:-1])
+                        radio.status_fn(status)
+                    else:
+                        raise errors.RadioError('Received bad ACK from Radio')
+
+    except Exception as e:
+        raise errors.RadioError(f"Exception During Download: {e}")
+    finally:
+        end_programming(serial)
+    return memmap.MemoryMapBytes(data)
+
+
+def write_data(serial, addr, data):
+    pfx = bytes.fromhex('57' + addr.zfill(8))
+    data = pfx + data
+    chk_int = sum(data) % 256
+    chk_byte = int.to_bytes(chk_int, 1, "little")
```

Comment:
`struct.pack()` here.. You can and should use it to pack the whole header in cases like this. It's much easier to reason about.